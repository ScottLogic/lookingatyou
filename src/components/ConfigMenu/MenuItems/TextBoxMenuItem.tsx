import React, { useState } from 'react';
interface ITextBoxMenuItemProps {
    name: string,
    isValidInput: { (text: string): boolean },
    onValidInput: { (text: string): void },
    defaultValue: string,
    parse: { (text: string): string }
}
export default function TextBoxMenuItem(props: ITextBoxMenuItemProps) {
    const [isValid, setIsValid] = useState(true);
    const [value, setValue] = useState(props.defaultValue);
    const [lastValidValue, setLastValidValue] = useState(props.defaultValue);
    return (
        <div>
            <label>{props.name}</label>
            <input type="textbox" value={value} style={{ color: isValid ? "black" : "red" }}
                onBlur={(event) => {
                    setValue(props.parse(lastValidValue));
                    setIsValid(true);
                }}
                onChange={(event) => {
                    var value = event.target.value;
                    setValue(value);
                    var isValid = props.isValidInput(event.target.value);
                    setIsValid(isValid);
                    if (isValid) {
                        setLastValidValue(value);
                        props.onValidInput(value)
                    }
                }}></input>
        </div>
    )
}