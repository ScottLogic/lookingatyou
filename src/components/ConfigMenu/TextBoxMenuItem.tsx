import React, { useState, useEffect } from 'react';
interface ITextBoxMenuItemProps {name: string, 
    isValidInput: { (text: string): boolean }, 
    onValidInput: { (text: string): void }, 
    value: string }
export default function TextBoxMenuItem(props: ITextBoxMenuItemProps) {
    const [isValid, setIsValid] = useState(true);
    return (
        <div>
            <label>{props.name}</label>
            <input type="textbox" defaultValue={props.value} style={{color : isValid ? "black" : "red"}} onChange={(event) => {
                var isValid = props.isValidInput(event.target.value);
                setIsValid(isValid);
                if (isValid)
                    props.onValidInput(event.target.value)
            }}></input>
        </div>
    )
}