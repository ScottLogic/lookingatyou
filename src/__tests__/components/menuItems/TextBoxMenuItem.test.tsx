import { mount, shallow } from 'enzyme';
import jsdom from 'jsdom';
import { wrap } from 'module';
import React from 'react';
import TextBoxMenuItem, {
    ITextBoxMenuItemProps,
} from '../../../components/configMenu/menuItems/TextBoxMenuItem';

let props: ITextBoxMenuItemProps;
let mockIsValidInput: jest.Mock;
let mockParse: jest.Mock;

describe('TextBoxMenuItem tests', () => {
    beforeEach(() => {
        mockIsValidInput = jest.fn();
        mockParse = jest.fn();
        props = {
            name: 'test',
            defaultValue: '30',
            isValidInput: mockIsValidInput,
            onValidInput: jest.fn(),
            parse: mockParse,
        };
    });

    it('should render correctly', () => {
        const wrapper = shallow(<TextBoxMenuItem {...props} />).debug();
        expect(wrapper).toMatchSnapshot();
    });

    it('should call isValidInput', () => {
        const value = '25';
        mockIsValidInput.mockReturnValue(false);
        const wrapper = mount(<TextBoxMenuItem {...props} />);
        wrapper.find('input').simulate('change', { target: value });
        expect(wrapper.get(0).props.isValidInput).toBeCalled();
    });

    it('should call onValidInput when input is valid', () => {
        const value = '1';
        mockIsValidInput.mockReturnValue(true);
        const wrapper = mount(<TextBoxMenuItem {...props} />);
        wrapper.find('input').simulate('change', { target: value });
        expect(wrapper.get(0).props.onValidInput).toBeCalled();
    });

    it('onBlur should reset input to last valid value when invalid value has been entered', () => {
        const value = '1';
        mockIsValidInput.mockReturnValue(false);
        mockParse.mockReturnValue(props.defaultValue);
        const wrapper = mount(<TextBoxMenuItem {...props} />);
        wrapper.find('input').simulate('focus');
        wrapper.find('input').simulate('change', { target: { value } });
        wrapper.find('input').simulate('blur');
        expect(wrapper.find('input').props().value).toBe(props.defaultValue);
    });
});
