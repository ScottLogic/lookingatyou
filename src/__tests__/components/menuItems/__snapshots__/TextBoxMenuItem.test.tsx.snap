// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBoxMenuItem tests should render correctly 1`] = `
"<div data-tip={true} data-for=\\"FPS\\">
  <label>
    test
  </label>
  <input type=\\"number\\" value={30} style={{...}} onBlur={[Function: onBlur]} onChange={[Function: onChange]} step={1} />
</div>"
`;
